pkg/monitoring/docker/docker.go

```diff
@@ -11,29 +12,47 @@ import (
 
 const dockerEndpointAddress = "unix:///var/run/docker.sock"
 
-var client *docker.Client
-var connectionSucceedOnce bool
+type DockerWatcher struct {
+	client                *docker.Client
+	connectionSucceedOnce bool
+}
 
-func ListContainers() (map[string]interface{}, error) {
-	var err error
-	if client == nil {
-		client, err = docker.NewClient(dockerEndpointAddress)
+func (dw *DockerWatcher) initClient() error {
+	if dw.client == nil {
+		_, err := docker.NewClient(dockerEndpointAddress)
 		if err != nil {
-			return nil, err
+			return err
 		}
 	}
+	return nil
+}
 
-	containers, err := client.ListContainers(docker.ListContainersOptions{All: false})
-	if err != nil && err == docker.ErrConnectionRefused && !connectionSucceedOnce {
-		// do not produce error if Docker is missing and wasn't successfully connected in this session before
+func (dw *DockerWatcher) ListContainers() (map[string]interface{}, error) {
+	err := dw.initClient()
```

Comment:
Is it for performance reason that we create the client so late? Otherwise we should init the client on creation of `DockerWatcher`

Reply:
the reason is that we don't have `New` method. If I put the client init only inside the `New()` it could be unclear why you can do `docker.New()` without params, but can't just use struct
```
d := docker.New()
d.ListContainers()
```

but can't do 
```
d := &DockerWatcher{}
d.ListContainers()
```

Reply:
> could be unclear why you can do `docker.New()` without params, but can't just use struct

_Exactly_ this is the reason. You should not use the struct directly except you do all the initialisation manually as well.
A comment in the code pointing this out should be fine... but I think it's sort of the norm that you have a `New` function in case the struct needs some init code

Reply:
done